cli/lock.ml

```diff
@@ -209,20 +171,107 @@ let get_pin_depends ~global_state local_opam_files =
   root_pin_depends local_opam_files >>= pull_pin_depends ~global_state
 
 let calculate_opam ~build_only ~allow_jbuilder ~local_opam_files ~ocaml_version
-    =
+    ~target_packages =
   let open Rresult.R.Infix in
   OpamGlobalState.with_ `Lock_none (fun global_state ->
       OpamSwitchState.with_ `Lock_none global_state (fun switch_state ->
           check_repo_config ~switch_state;
           get_pin_depends ~global_state local_opam_files >>= fun pin_depends ->
           Opam_solve.calculate ~build_only ~allow_jbuilder ~local_opam_files
-            ~pin_depends ?ocaml_version switch_state))
+            ~target_packages ~pin_depends ?ocaml_version switch_state))
+
+let subselect_explicit ~local_packages ~explicit_list =
+  let by_opam_name entry (explicit_name, _) =
+    let name = entry |> Package_argument.name |> OpamPackage.Name.to_string in
+    String.equal name explicit_name
+  in
+
+  List.fold_left
+    ~f:(fun acc explicit ->
+      match (List.find_opt ~f:(by_opam_name explicit) local_packages, acc) with
+      | None, Ok _ -> Error [ explicit ]
+      | None, Error errors -> Error (explicit :: errors)
+      | Some _, Error errors -> Error errors
+      | Some x, Ok xs -> Ok (x :: xs))
+    ~init:(Ok []) explicit_list
+  |> Result.map_error ~f:(fun missing_packages ->
+         let msg =
+           Fmt.str "Packages %a specified but not found in repository"
+             Fmt.(list ~sep:comma Package_argument.pp_styled)
+             missing_packages
+         in
+         `Msg msg)
+
+let only_return_root xs =
+  let root = Sys.getcwd () |> Fpath.v in
```

Comment:
We shouldn't use `getcwd` here but rather use the `repo` argument. It defaults to `cwd` but can be overwritten using a CLI option.